### DISCUSSION TITLE
Why are there still some "temporary variable" involved?

### DISCUSSION THREAD

**Discussion by user202729** [lvjr/functional#13] [General]
Consider `tlVarReplaceAll` for example.

Why isn't it simply `tlReplaceAll {tl} {old} {new}`, then the user can easily do the `tlVarReplaceAll` if they need with a `\tlSet \var {\tlReplaceAll { \expValue \var } {old} {new}}`?

Or even -- `\varSet \var { \convertToStr {...} }` instead of `\strSet \var {...}`?

Sure, that's more verbose and slower, but the aim here is to make it *intuitive* for users, right? I think it's better to introduce the most general tools to users, and teach them how to build specific ones.

The expl3 interface was only made that way because not all operations are expandable. In this package there's no such limitation.

------

On another similar note, instead of having e.g. both `\tlLog` and `\tlVarLog`, it might be better to have only `\tlLog` and the other can be done with `\tlLog{\expVar{...}}`. Or move the "induced" commands to a separate "command shorthand" section -- they're not as important, can be discussed later.

**Comment by lvjr** [comment]
You may compare them with syntactic sugar in other programming languages.They could be removed, but there is no harm in providing them and sometime they are useful.

**Reply by user202729** [reply]
I don't think so, this is the opposite of syntactic sugar in that it makes the syntax more verbose.

For example if I want to do a tl replacement on a variable, then pass it to another function while keeping the original value of the variable

```
\tlSet \varA  {123}
\tlSetEq \tmp \varA
\tlVarReplaceAll \tmp {old} {new}
\function {\tlUse \tmp}
```

compared to


```
\tlSet \varA  {123}
\function {\tlReplaceAll {\tlUse\varA} {old} {new}}
```

?
(if we use Python as a comparison, all of its functions are of the second type; but to be fair Python doesn't have pass-variable-by-reference)

**Reply by lvjr** [reply]
Sorry, I can't see any problems here.

**Reply by user202729** [reply]
I'm saying I have to use a temp variable and have to come up with some meaningful name for it, even though most of the time there isn't any meaningful name.

Sometimes using temp variable is a good thing (e.g. to break down a complex expression), but when you can't think of a meaningful name of it it's probably bad.

(for comparison: the way TeX programming works means you need to define lots of internal temp macros to do work, and most of the time they aren't named meaningfully either. In source3 the problem is less serious but not completely gone, search for all the aux, auxi, auxii occurrences in expl3 source code.)

**Reply by lvjr** [reply]
You can use `\Functional{scoping=true}`, see the manual.

**Reply by user202729** [reply]
I'm aware of the feature, just saying that in my opinion having meaningless variable names around makes the *source code* (not the runtime environment) cluttered.

For comparison, Python has `re.sub()` and `<string object>.replace()` returns a value, Lua has `string.gsub()` returns a value.

**Reply by lvjr** [reply]
Oh, finally I got your point. I think you only need to say `\tlReplaceAll` is still missing in `functional` package. That is because `expl3` doesn't provide it, and I have to write it from scratch. Obviously I have to add it and its friends some day. Recently I don't have much time to spend on TeX. But if you would like to help to implement them, I will merge the code.

**Reply by user202729** [reply]
You don't need to write from scratch...? Just do something like

```
\prgNewFunction \tlReplaceAll {mmm} {
\tlSet \tmp {#1}
\tlVarReplaceAll \tmp {#2} {#3}
\return {\tlUse \tmp}
}
```

or something like that.

The point is, I think one of the powers of `functional` package is to allow you to nest unexpandable function inside another function, so it's a waste to not use them (in normal TeX, code only fully expandable functions can conveniently "return value" to other function as I mentioned once before).

Or make it more general for a "template code writer" to shorten the code something like

```
\prgNewFunction \varIOToNormalFunctionThreeArg {mm} { % first arg is var function name, second arg is normal function name
\prgNewFunction #2 {mmm} {
\tlSet \tmp {##1}
#1 \tmp ...
\return {\tlUse \tmp}
}
}
\varIOToNormalFunctionThreeArg \tlVarReplaceAll \tlReplaceAll
\varIOToNormalFunctionThreeArg \tlVarReplaceOnce \tlReplaceOnce
\varIOToNormalFunctionThreeArg \strVarReplaceAll \strReplaceAll
% etc., there are many such functions in expl3.
```
